cinder/backup/chunkeddriver.py

```diff
@@ -780,3 +776,161 @@ def delete_backup(self, backup):
+
+        If there is no further segment for the same object, we close and
+        remove the corresponding object reader, freeing up the memory.
+        """
+        obj_name = segment.obj['name']
+        for _segment in self._segments[self._idx:]:
+            if obj_name == _segment.obj['name']:
+                return
+
+        obj_reader = self._object_readers[obj_name]
+        if obj_reader:
+            obj_reader.close()
+            self._object_readers.pop(obj_name)
+
+    def add_object(self, metadata_object):
+        """Merges a backup chunk over the self._segments list.
+
+        The backup chunks are expected to come in order.
+        :param metadata_object: the backup chunk
+        """
+        # make a copy because we will modify it later
+        alt_obj = metadata_object.copy()
+        found = False
+        for idx, segment in enumerate(self._segments):
+            offset = alt_obj['offset']
+            length = alt_obj['length']
+            end = offset + length
+            alt_idx = idx
+
+            # the object can be merged with this segment
+            if segment.offset <= offset < segment.end:
+                found = True
+                # remove the segment from the list, we're going to re-add
+                # only parts of it to the list or nothing at all
+                self._segments.pop(idx)
+
+                # if the object starts after this segment's offset, then we
+                # keep the beginning of this segment
+                diff = offset - segment.offset
+                if diff > 0:
+                    self._segments.insert(alt_idx,
+                                          Segment.of(segment,
+                                                     length=diff))
+                    alt_idx += 1
+
+                # if the object ends before this segment's end, then we keep
+                # the last part of this segment, otherwise we don't
+                diff = segment.end - end
+                if diff > 0:
+                    self._segments.insert(alt_idx, Segment(alt_obj))
+                    alt_idx += 1
+                    self._segments.insert(alt_idx,
+                                          Segment.of(segment,
+                                                     length=diff,
+                                                     offset=end))
+                else:
+                    self._segments.insert(alt_idx,
+                                          Segment(alt_obj,
+                                                  length=length + diff))
+                    # if there is nothing left from this object, we're done
+                    if diff == 0:
+                        break
+                    # if there is something left from this object, continue
+                    # merging it over the next segment
+                    alt_obj['offset'] = segment.end
+                    alt_obj['length'] = abs(diff)
+
```

Comment:
I have a hard time understanding this part, as the `enumerate(self._segments)` while we edit `self._segments` kills my thinking process. I've written a small script to play around with this:
```
segments = ['a', 'b', 'c', 'd', 'e']

for i, s in enumerate(segments):
  print(i, s)
  if s == 'b':
    old_i = i
    segments.pop(i)
    segments.insert(old_i, 'f')
    old_i += 1
    segments.insert(old_i, 'g')
  if s == 'c':
    old_i = i
    segments.pop(i)
    segments.insert(old_i, 'h')
    old_i += 1
    segments.insert(old_i, 'i')

print(segments)
```
Output:
```
$ python /tmp/bla.py 
(0, 'a')
(1, 'b')
(2, 'g')
(3, 'c')
(4, 'i')
(5, 'd')
(6, 'e')
['a', 'f', 'g', 'h', 'i', 'd', 'e']
```
From this I gather, that we can first find a matching segment, pop that segment, add a part of that segment again, add ourselves […]. In the next iteration, we will see ourselves, pop ourselves, see no diff, add ourselves again and break, because diff == 0. That's really confusing to me. Can we have a faster code-path for this at the beginning of the for-lop with a comment?  `if segment.obj.['name'] == alt_obj['name']: continue` or something, explaining that with our shenanigans of iterating over a list we change we will end up seeing us added to the list?

I still don't think I fully grasp this with all the different possibilities.

Reply:
> ... In the next iteration, we will see ourselves ...

Do you mean the next iteration of the same `for` loop? 
If yes, that's not the case. We will not see ourselves again in the next iteration.

````
segments = ['a', 'b', 'c']
ourselves = 'f'
for i, s in enumerate(segments):
    print(s)
    if s == 'b':
        segments.pop(i)
        segments.insert(i, ourselves)
    if s == ourselves:
        print('we are seeing ourselves again')
````
Output:
````
a
b
c
````

Reply:
I get it.
I changed that to a while so that it can jump over what was already inserted.